src/Cooked/MockChain/Testing.hs

```diff
@@ -295,10 +133,175 @@ assertSubset l r =
         l
     )
 
+-- | Asserts whether 2 sets are equal, both given as lists.
 assertSameSets :: (Show a, Eq a) => [a] -> [a] -> HU.Assertion
 assertSameSets l r =
   HU.assertBool
     ("expected lists of the same length, got " ++ show (length l) ++ " and " ++ show (length r))
     (length l == length r)
     .&&. assertSubset l r
     .&&. assertSubset r l
+
+-- * Testing mockchain runs
+
+-- | Data structure to test a mockchain run
+data Test a prop = Test
+  { -- | The mockchain run to test
+    testRun :: StagedMockChain a,
+    -- | The initial distribution from which the run should be ran
+    testInitDist :: InitialDistribution,
+    -- | The property should hold in case of failure
+    testErrorProp :: PrettyCookedOpts -> MockChainError -> [MockChainLogEntry] -> prop,
+    -- | The property that should hold in case of success
+    testResultProp :: PrettyCookedOpts -> a -> UtxoState -> [MockChainLogEntry] -> prop,
```

Comment:
Do we agree that this will eventually be changed once we allow continuation after transaction failure?

I agree this is useful now but I'd like to be certain we are on the same page.

After we allow traces to continue after tx failure, there won't be any concept of successful traces / failed traces. Instead, we will always have a `UtxoState` at the end, and `MockChainError`s will be stored as log entries in the journal `[MockChainLogEntry]`.

So most likely, in the future, we will have a single `testProp :: PrettyCookedOpts -> [MockChainLogEntry] -> UtxoState -> prop` right? Along with smart constructors of course (e.g. for "there exist a phase 2 failure in the journal", etc).

Reply:
I don't completely agree, because there will still be other possible cases of failures outside of validation failure. All other mockchain errors will still result in a failure. Only validation errors could be handled as part as predicate over the journal.

Reply:
Oh yes that's right. Thanks.